#98 Changed backend

MaximSmolskiy opened the pull request.

MaximSmolskiy committed:
#98 Changed packages

MaximSmolskiy committed:
#98 Deleted class Colors

MaximSmolskiy committed:
#98 Added class Cell

MaximSmolskiy committed:
#98 Added class StateMachine

MaximSmolskiy committed:
#98 Changed class NonogramSolver

MaximSmolskiy committed:
#98 Changed class Field

MaximSmolskiy committed:
#98 Changed class Nonogram

MaximSmolskiy committed:
#98 Changed class ImageConverter

MaximSmolskiy committed:
#98 Updated class Cell

- Added constants for directions

MaximSmolskiy committed:
#98 Updated class Field

- Returned Parcelable
- Returned method getAnotherColorIndex
- Returned method getColor
- Returned method setColor

MaximSmolskiy committed:
#98 Updated class NonogramSolver

- Updated method solve
- Updated method init

MaximSmolskiy committed:
#98 Updated method checkCorrectness

MaximSmolskiy committed:
#98 Updated AsyncTaskConvertImage

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/ImageConverter.java:
```diff
@@ -26,25 +27,26 @@
         rows = min(rows, bmp.getHeight());
         cols = min(cols, bmp.getWidth());
         Bitmap bw = getBlackWhite(bmp.copy(bmp.getConfig(), bmp.isMutable()));
         Field field = null;
```
Replace `Field` by `Nonogram`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/ImageConverter.java:
```diff
@@ -26,25 +27,26 @@
+        solver.setAsyncTask(asyncTask);
         for (int p = 128; !asyncTask.isCancelled() && p <= 248; p += 5) {
-            field = getThumbnail(bw, rows, cols, p);
+            field = new Field(bw, rows, cols, p);
```
Replace `Field` by `Nonogram`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/ImageConverter.java:
```diff
@@ -26,25 +27,26 @@
         for (int p = 128; !asyncTask.isCancelled() && p <= 248; p += 5) {
-            field = getThumbnail(bw, rows, cols, p);
+            field = new Field(bw, rows, cols, p);
             solver.setNonogram(new Nonogram(field));
```
Don't create a new instance.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/ImageConverter.java:
```diff
@@ -26,25 +27,26 @@
+            field = new Field(bw, rows, cols, p);
             solver.setNonogram(new Nonogram(field));
-            if (solver.solve()) return field;
+            if (solver.solve()) return new Nonogram(field);
```
Return instance without a constructor.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/ImageConverter.java:
```diff
@@ -26,25 +27,26 @@
          *TODO: `field` needs to be replaced by `null` when solver will be ready
          */
-        return field;
+        return new Nonogram(field);
```
Return instance without a constructor.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,324 @@
+     * Private constructor without parameters
+     */
+    private Field() {
+
```
Optional line.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,324 @@
+                for (int x = j * w; x < (j + 1) * w; x++) {
+                    for (int y = i * h; y < (i + 1) * h; y++) {
+                        int color = bmp.getPixel(x, y);
+                        if (color == WHITE) {
```
It can be replaced by
```java
                        if (bmp.getPixel(x, y) == WHITE) sumColors += 255;
```

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,324 @@
+                    }
+                }
+                int avgColor = sumColors / (w * h);
+                if (avgColor <= p) {
```
It can be replaced by
```java
                if (sumColors <= p * w * h) {
```

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,324 @@
+        int h = height / rows, w = width / cols;
+        for (int i = 0; i < rows; i++) {
+            for (int j = 0; j < cols; j++) {
+                int sumColors = 0;
```
This variable can be declared before a loop.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,324 @@
+     * @return index of first cell with another color in row or column
+     */
+    public int getAnotherColorIndex(int ind, int pos, int dir, int type) {
+        int i = pos + 1;
```
Now it can be replaced by
```java
        int i = pos + 1, len = getLength(type);
        for (; i >= 0 && i < len; i += dir) {
            if (getColor(ind, i, type) != getColor(ind, pos, type)) break;
        }
        return i;
```

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,324 @@
+     * @return thumbnail
+     */
+    Bitmap getBitmap() {
+        Field field = translateToColors();
```
It could be better if it will be replaced by
```java
        Field field = this;
        if (_state == INDEX)
            field = translateToColors();
```

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,324 @@
+        Field field = translateToColors();
+        Bitmap bmp = Bitmap.createBitmap(_cols,
+                _rows,
+                Bitmap.Config.ARGB_8888);
```
Is it possible to change it on `Bitmap.Config.RGB_565`?

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,324 @@
+        field._state = COLOR;
+        field._rows = _rows;
+        field._cols = _cols;
+        field._field = _field;
```
It doesn't copy. You can use next:
```java
        field._field = new int[_rows][];
        for (int i = 0; i < _rows; ++i) {
            field._field[i] = _field[i].clone();
        }
```

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,324 @@
+        if(_state == INDEX) {
+            for (int i = 0; i < field._rows; i++) {
+                for (int j = 0; j < field._cols; j++) {
+                    field._field[i][j] = _colors[_field[i][j]];
```
There is no guarantee that `_field` doesn't contain cells with value `EMPTY`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,324 @@
+     *
+     * @return count of colors
+     */
+   public int getColors() {
```
There are four spaces before `public`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,324 @@
+     *
+     * @return thumbnail
+     */
+    Bitmap getBitmap() {
```
Make it `public`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,296 @@
+    /**
+     * Field of the puzzle
+     */
+    Field _field;
```
Make it `private`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,296 @@
+     *
+     * @param fieldNono - nonogram field
+     */
+    public Nonogram(@NonNull Field fieldNono) {
```
Make it `private`

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,296 @@
+     * Private constructor without parameters
+     */
+    private Nonogram() {
+
```
Optional line.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,296 @@
+     */
+    public boolean checkCorrectness() {
+        int rows = _field.getRows(), cols = _field.getCols();
+        for (int i = 0; i < rows; i++) {
```
Is it possible to clone `_field`?

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,296 @@
+                }
+            }
+        }
+        return isEqual(new Nonogram(_field));
```
Is it possible to not create a new instance of the class `Nonogram` to check correctness?

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,296 @@
+    /**
+     * Method for clearing field of the puzzle
+     */
+    void clearField() {
```
Make it `public`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,296 @@
+     *
+     * @return new nonogram that contains field with colors in cells
+     */
+    Nonogram translateToColors() {
```
Make it `public`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,296 @@
+            for (int j = 0, k, length = 0; j < inLim; ) {
+                k = field.getAnotherColorIndex(i, j, 1, type);
+                int color = field.getColor(i, j, type);
+                if (color == Field.BLACK) {
```
It can be replaced by 
```java
                if (field.getColor(i, j, type) == Field.BLACK) {
```

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/ui/editimage/EditImageActivity.java:
```diff
@@ -23,8 +23,9 @@
 
 import com.mithridat.nonoconverter.R;
-import com.mithridat.nonoconverter.backend.Field;
+import com.mithridat.nonoconverter.backend.nonogram.Field;
```
Unused import statement.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/solver/Cell.java:
```diff
@@ -0,0 +1,63 @@
+    /**
+     * Diagonal direction
+     */
+    public static final int DIAG = 0;
```
Make it `package-private`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java:
```diff
@@ -76,59 +70,32 @@ public void setAsyncTask(AsyncTask<Void, Void, Field> asyncTask) {
-            isChanged = applyAlgorithms(_cols, Field.COL) || isChanged;
+            if (!applyAlgorithmFSM(_cols, Field.COL)) return false;
             _cols.clear();
+            if(_nono.checkCorrectness()) return true;
```
Replace it by
```java
            if(_rows.isEmpty()) break;
```

maik-nack replied:
Sorry. Add space after `if`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,324 @@
+     */
+    Bitmap getBitmap() {
+        Field field = translateToColors();
+        Bitmap bmp = Bitmap.createBitmap(_cols,
```
Check [Coding standard](https://github.com/nonoteam/NonoConverter/blob/master/Development.md#line-wrapping-strategies).

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/solver/Cell.java:
```diff
@@ -0,0 +1,63 @@
+    /**
+     * Horizontal direction
+     */
+    public static final int HOR = 1;
```
Make it `package-private`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java:
```diff
@@ -2,35 +2,25 @@
 /**
  * Nonogram solving class
  */
 public class NonogramSolver {
```
Is it possible to make all fields `private` and create getters for all except `AsyncTask`?

maik-nack replied:
You can make getters `package-private` if you want.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java:
```diff
@@ -76,59 +70,32 @@ public void setAsyncTask(AsyncTask<Void, Void, Field> asyncTask) {
+        _rowsFSM = new StateMachine[rows];
+        initFSM(_rowsFSM, Field.ROW);
+        _colsFSM = new StateMachine[cols];
+        initFSM(_colsFSM, Field.COL);
```
It could be better to move lines 95-98 to `setNonogram`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java:
```diff
@@ -145,139 +112,46 @@ boolean applySimpleBoxes(int ind, int type) {
-            Arrays.fill(array[i], false);
+    private void applyAlgorithmSB(int type) {
+        int length = _nono.getFirstLength(type);
+        for (int i = 0; i < length; i++) {
```
Do we need a check of cancellation of the `AsyncTask` here?

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,296 @@
+            int type,
+            int outLim,
+            int inLim,
+            Field field) {
```
Now we don't need to pass `Field` in this function.

MaximSmolskiy committed:
#98 Updated class ImageConverter

- Fixed issues from review

MaximSmolskiy committed:
#98 Updated class Cell

- Fixed issues from review

MaximSmolskiy committed:
#98 Updated class EditImageActivity

- Deleted unused import statement

MaximSmolskiy committed:
#98 Updated class Field

- Fixed issues from review

MaximSmolskiy committed:
#98 Updated class NonogramSolver

- Fixed issues from review

MaximSmolskiy committed:
#98 Updated class Nonogram

- Fixed issues from review

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,339 @@
+     * @return index of first cell with another color in row or column
+     */
+    public int getAnotherColorIndex(int ind, int pos, int dir, int type) {
+        int i = pos + 1, len = getLength(ROW + COL - type);
```
Why `ROW + COL - type`?

MaximSmolskiy replied:
Because it was: `len = _cols` if `type == ROW` and `len = _rows` if `type == COL`.

maik-nack replied:
Ok. Is it possible to change the semantic of the method `getLength` or create a new method? New semantic: if the type is `ROW` method returns a count of columns i.e. length of a row.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,339 @@
+                    if (_field[i][j] != EMPTY) {
+                        field._field[i][j] = _colors[_field[i][j]];
+                    } else {
+                        field._field[i][j] = EMPTY;
```
Color for `EMPTY` is `WHITE`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,339 @@
+    }
+
+    public Field(Field other) {
+        _colors = other._colors;
```
It doesn't copy. You can use next:
```java
        _colors = other._colors.clone();
```

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/solver/NonogramSolver.java:
```diff
@@ -2,53 +2,91 @@
 import java.util.HashSet;
 
+import static com.mithridat.nonoconverter.backend.nonogram.Field.ROW;
+
```
You can also import `Field.COL`.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,339 @@
+        }
+    }
+
+    public Field(Field other) {
```
This method usually calls `clone` in Java.

maik-nack replied:
Sorry, the semantic of the `clone` is some another. You can rename it to `copy` if you want.

MikhailNakhatovich commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -256,8 +260,8 @@ private void initGrid() {
 
-        fillNonogramGrid(_left, Field.ROW, h, w, _field);
-        fillNonogramGrid(_top, Field.COL, w, h, _field);
+        fillNonogramGrid(_left, Field.ROW, h, w);
```
You can use `ROW` here.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,330 @@
+     * @return length of _left or _top
+     */
+    public int getFirstLength(int type) {
+        if (type == Field.ROW) return getLeftRowsLength();
```
You can use `ROW` here.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,330 @@
+     * @return length of _left[i] or _top[i]
+     */
+    public int getSecondLength(int i, int type) {
+        if (type == Field.ROW) return getLeftRowLength(i);
```
You can use `ROW` here.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,330 @@
+     * @return _left[i][j] or _top[i][j]
+     */
+    public int getValue(int i, int j, int type) {
+        if (type == Field.ROW) return getLeftRowValue(i, j);
```
You can use `ROW` here.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,330 @@
+        _left = new int[h][];
+        _top = new int[w][];
+
+        fillNonogramGrid(_left, Field.ROW, h, w);
```
You can use `ROW` here.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/ui/editimage/EditImageActivity.java:
```diff
@@ -606,11 +606,11 @@ protected Field doInBackground(Void... params) {
             _activity._pdLoading.dismiss();
             Intent intent = new Intent(_activity, ResultActivity.class);
-            intent.putExtra(ActivitiesConstants.EX_NONO_FIELD, result);
+            intent.putExtra(ActivitiesConstants.EX_NONO_FIELD, result.getField());
```
The length of this line over than 80.

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,339 @@
+    Field translateToColors() {
+        Field field = new Field(this);
+        if(_state == INDEX) {
+            for (int i = 0; i < field._rows; i++) {
```
Please add a line here
```java
            field._state = COLOR;
```

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,339 @@
+        Bitmap bmp = Bitmap.createBitmap(_cols, _rows, Bitmap.Config.RGB_565);
+        for (int i = 0; i < _rows; i++) {
+            for (int j = 0; j < _cols; j++) {
+                bmp.setPixel(j, i, field._field[j][i]);
```
It needs
```java
                bmp.setPixel(j, i, field._field[i][j]);
```

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Field.java:
```diff
@@ -0,0 +1,339 @@
+        _state = INDEX;
+        _rows = rows;
+        _cols = cols;
+        _field = new int[rows][cols];
```
It needs to sort `_colors` and fill `_field` by the index of `WHITE`.

MaximSmolskiy committed:
#98 Fixed line with length over than 80

MaximSmolskiy committed:
#98 Updated class NonogramSolver

- Changed Field.COL by COL
- Made getters package-private
- Added TODO

MaximSmolskiy committed:
#98 Updated class ImageConverter

- Added resizing

MaximSmolskiy committed:
#98 Updated class Field

- Fixed bugs
- Changed copy constructor to method copy
- Changed semantics of method getLength

MaximSmolskiy committed:
#98 Updated class Nonogram

- Changed Field.COL by COL
- Changed Field.EMPTY by EMPTY
- Changed Field.BLACK by BLACK

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,332 @@
+            grid[i] = new int[0];
+            for (int j = 0, k, length = 0; j < inLim; ) {
+                k = _field.getAnotherColorIndex(i, j, 1, type);
+                if (_field.getColor(i, j, type) ==
```
`==` is an operator. Check [Coding standard](https://github.com/nonoteam/NonoConverter/blob/master/Development.md#line-wrapping-strategies).

maik-nack commented on app/src/main/java/com/mithridat/nonoconverter/backend/nonogram/Nonogram.java:
```diff
@@ -0,0 +1,332 @@
+        for (int i = 0; i < rows; i++) {
+            for (int j = 0; j < cols; j++) {
+                if (field.getColor(i, j) == EMPTY) {
+                    field.setColor(i, j, 0);
```
Maybe zero needs to be replaced by `_field.getColorState(WHITE)`.

MaximSmolskiy committed:
#98 Updated class Nonogram

- Replaced zero by field.getColorState(WHITE)
- Fixed coding style error

maik-nack reviewed (approved):
GREAT JOB!!!

NikitaSukhanov reviewed (approved):
Looks great!
